panther_battery/include/panther_battery/battery.hpp

```diff
@@ -0,0 +1,105 @@
+#ifndef PANTHER_BATTERY_BATTERY_HPP_
+#define PANTHER_BATTERY_BATTERY_HPP_
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <string_view>
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <sensor_msgs/msg/battery_state.hpp>
+
+#include <panther_utils/moving_average.hpp>
+
+namespace panther_battery
+{
+
+using BatteryStateMsg = sensor_msgs::msg::BatteryState;
+
+struct BatteryParams
+{
+  const int64_t voltage_window_len;
+  const int64_t temp_window_len;
+  const int64_t current_window_len;
+  const int64_t charge_window_len;
+};
+
+class Battery
+{
+public:
+  Battery(
+    std::function<float()> read_voltage, std::function<float()> read_current,
+    std::function<float()> read_temp, std::function<float()> read_charge,
+    const BatteryParams & params);
+
+  ~Battery() {}
+
+  bool Present();
+  void Update(rclcpp::Time & header_stamp, const bool charger_connected);
+  void Reset(rclcpp::Time & header_stamp);
+  bool HasErrorMsg() const;
+
+  std::string GetErrorMsg() const;
+  BatteryStateMsg GetBatteryMsg() const;
+  BatteryStateMsg GetBatteryMsgRaw() const;
+
+private:
+  float ADCToBatteryVoltage(const float adc_data);
+  float ADCToBatteryCurrent(const float adc_data);
+  float ADCToBatteryCharge(const float adc_data);
+  float ADCToBatteryTemp(const float adc_data);
+  float ADCToBatteryVoltageTemp(const float adc_data);
```

Comment:
```suggestion
  inline float ADCToBatteryVoltage(const float adc_data);
  inline float ADCToBatteryCurrent(const float adc_data);
  inline float ADCToBatteryCharge(const float adc_data);
  float ADCToBatteryTemp(const float adc_data);
  inline float ADCToBatteryVoltageTemp(const float adc_data);
```
`ADCToBatteryTemp` was left not inline intentionally